File: app/src/main/java/tech/pauly/findapet/discover/DiscoverViewModel.java
package tech.pauly.findapet.discover;

import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.OnLifecycleEvent;
import android.databinding.ObservableArrayList;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;
import android.databinding.ObservableInt;
import android.databinding.ObservableList;
import android.location.Address;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import tech.pauly.findapet.R;
import tech.pauly.findapet.data.AnimalRepository;
import tech.pauly.findapet.data.FilterRepository;
import tech.pauly.findapet.data.PetfinderException;
import tech.pauly.findapet.data.models.Age;
import tech.pauly.findapet.data.models.Animal;
import tech.pauly.findapet.data.models.AnimalListResponse;
import tech.pauly.findapet.data.models.AnimalSize;
import tech.pauly.findapet.data.models.AnimalType;
import tech.pauly.findapet.data.models.FetchAnimalsRequest;
import tech.pauly.findapet.data.models.Filter;
import tech.pauly.findapet.data.models.Sex;
import tech.pauly.findapet.shared.BaseViewModel;
import tech.pauly.findapet.shared.LocationHelper;
import tech.pauly.findapet.shared.PermissionHelper;
import tech.pauly.findapet.shared.ResourceProvider;
import tech.pauly.findapet.shared.datastore.DiscoverAnimalTypeUseCase;
import tech.pauly.findapet.shared.datastore.DiscoverToolbarTitleUseCase;
import tech.pauly.findapet.shared.datastore.FilterUpdatedUseCase;
import tech.pauly.findapet.shared.datastore.TransientDataStore;
import tech.pauly.findapet.shared.events.PermissionEvent;
import tech.pauly.findapet.shared.events.PermissionListener;
import tech.pauly.findapet.shared.events.ViewEventBus;

import static android.Manifest.permission.ACCESS_FINE_LOCATION;

public class DiscoverViewModel extends BaseViewModel {

    public ObservableInt columnCount = new ObservableInt(2);
    public ObservableBoolean refreshing = new ObservableBoolean(false);
    public ObservableBoolean locationMissing = new ObservableBoolean(false);
    public ObservableBoolean animalsMissing = new ObservableBoolean(false);
    public ObservableList<Chip> chipList = new ObservableArrayList<>();
    public ObservableField<Chip> locationChip = new ObservableField<>();

    private final AnimalListAdapter listAdapter;
    private final AnimalListItemViewModel.Factory animalListItemFactory;
    private final AnimalRepository animalRepository;
    private TransientDataStore dataStore;
    private PermissionHelper permissionHelper;
    private ViewEventBus eventBus;
    private LocationHelper locationHelper;
    private ResourceProvider resourceProvider;
    private FilterRepository filterRepository;

    private AnimalType animalType = AnimalType.CAT;
    private int lastOffset = 0;
    private boolean firstLoad = true;

    @Inject
    public DiscoverViewModel(AnimalListAdapter listAdapter,
                             AnimalListItemViewModel.Factory animalListItemFactory,
                             AnimalRepository animalRepository,
                             TransientDataStore dataStore,
                             PermissionHelper permissionHelper,
                             ViewEventBus eventBus,
                             LocationHelper locationHelper,
                             ResourceProvider resourceProvider,
                             FilterRepository filterRepository) {
        this.listAdapter = listAdapter;
        this.animalListItemFactory = animalListItemFactory;
        this.animalRepository = animalRepository;
        this.dataStore = dataStore;
        this.permissionHelper = permissionHelper;
        this.eventBus = eventBus;
        this.locationHelper = locationHelper;
        this.resourceProvider = resourceProvider;
        this.filterRepository = filterRepository;

        DiscoverAnimalTypeUseCase useCase = dataStore.get(DiscoverAnimalTypeUseCase.class);
        if (useCase != null) {
            animalType = useCase.getAnimalType();
        }
    }

    public AnimalListAdapter getListAdapter() {
        return listAdapter;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    public void onResume() {
        FilterUpdatedUseCase useCase = dataStore.get(FilterUpdatedUseCase.class);
        if (firstLoad || useCase != null) {
            firstLoad = false;
            requestPermissionToLoad();
        }
    }

    public void requestPermissionToLoad() {
        if (permissionHelper.hasPermissions(ACCESS_FINE_LOCATION)) {
            loadList();
        } else {
            eventBus.send(new PermissionEvent(this,
                                              new String[] { ACCESS_FINE_LOCATION },
                                              locationPermissionResponseListener(),
                                              100));
        }
    }

    private void loadList() {
        dataStore.save(new DiscoverToolbarTitleUseCase(animalType.getToolbarName()));
        listAdapter.clearAnimalItems();
        lastOffset = 0;
        fetchAnimals();
    }

    public void loadMoreAnimals() {
        fetchAnimals();
    }

    public boolean getErrorVisible() {
        return locationMissing.get() || animalsMissing.get();
    }

    private void fetchAnimals() {
        refreshing.set(true);
        onLifecycle(Observable.zip(getCurrentLocation(),
                                   getCurrentFilter(),
                                   (location, filter) -> new FetchAnimalsRequest(animalType, lastOffset, location.getPostalCode(), filter))
                              .flatMap(animalRepository::fetchAnimals)
                              .subscribe(this::setAnimalList, this::showError));
    }

    private Observable<Filter> getCurrentFilter() {
        return filterRepository.getCurrentFilterAndNoFilterIfEmpty()
                               .doOnSubscribe(this::removeFilterChips)
                               .doOnSuccess(this::addFilterChips).toObservable();
    }

    private Observable<Address> getCurrentLocation() {
        return locationHelper.fetchCurrentLocation()
                             .doOnSubscribe(this::removeLocationChip)
                             .doOnNext(this::addLocationChip);
    }

    private void removeFilterChips(Disposable disposable) {
        chipList.clear();
    }

    private void addFilterChips(Filter filter) {
        if (filter.getSex() != Sex.MISSING) {
            chipList.add(new Chip(resourceProvider.getString(filter.getSex().getFormattedName())));
        }

        if (filter.getAge() != Age.MISSING) {
            chipList.add(new Chip(resourceProvider.getString(filter.getAge().getFormattedName())));
        }

        if (filter.getSize() != AnimalSize.MISSING) {
            chipList.add(new Chip(resourceProvider.getString(filter.getSize().getFormattedName())));
        }

        if (!filter.getBreed().equals("")) {
            chipList.add(new Chip(filter.getBreed()));
        }
    }

    private void removeLocationChip(Disposable disposable) {
        locationChip.set(null);
    }

    private void addLocationChip(Address location) {
        locationChip.set(new Chip(resourceProvider.getString(R.string.chip_near_location, location.getPostalCode())));
    }

    private void showError(Throwable throwable) {
        refreshing.set(false);
        if (throwable instanceof PetfinderException) {
            switch (((PetfinderException) throwable).getStatusCode()) {
                case ERR_NO_ANIMALS:
                    animalsMissing.set(true);
                    return;
            }
        }
        throwable.printStackTrace();
    }

    private void setAnimalList(AnimalListResponse animalListResponse) {
        refreshing.set(false);
        List<AnimalListItemViewModel> viewModelList = new ArrayList<>();
        animalsMissing.set(false);
        lastOffset = animalListResponse.getLastOffset();
        for (Animal animal : animalListResponse.getAnimalList()) {
            viewModelList.add(animalListItemFactory.newInstance(animal));
        }
        listAdapter.setAnimalItems(viewModelList);
    }

    private PermissionListener locationPermissionResponseListener() {
        return response -> {
            if (response.getPermission().equals(ACCESS_FINE_LOCATION)) {
                if (response.isGranted()) {
                    locationMissing.set(false);
                    requestPermissionToLoad();
                } else {
                    locationMissing.set(true);
                }
            }
        };
    }
}

File: app/src/main/java/tech/pauly/findapet/discover/DiscoverViewModel.kt
package tech.pauly.findapet.discover

import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.arch.lifecycle.Lifecycle
import android.arch.lifecycle.OnLifecycleEvent
import android.databinding.*
import android.location.Address
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import tech.pauly.findapet.R
import tech.pauly.findapet.data.AnimalRepository
import tech.pauly.findapet.data.FilterRepository
import tech.pauly.findapet.data.PetfinderException
import tech.pauly.findapet.data.models.*
import tech.pauly.findapet.shared.*
import tech.pauly.findapet.shared.datastore.DiscoverAnimalTypeUseCase
import tech.pauly.findapet.shared.datastore.DiscoverToolbarTitleUseCase
import tech.pauly.findapet.shared.datastore.FilterUpdatedUseCase
import tech.pauly.findapet.shared.datastore.TransientDataStore
import tech.pauly.findapet.shared.events.*
import tech.pauly.findapet.utils.Chip
import java.util.*
import javax.inject.Inject

class DiscoverViewModel @Inject
constructor(val listAdapter: AnimalListAdapter,
            private val animalListItemFactory: AnimalListItemViewModel.Factory,
            private val animalRepository: AnimalRepository,
            private val dataStore: TransientDataStore,
            private val permissionHelper: PermissionHelper,
            private val eventBus: ViewEventBus,
            private val locationHelper: LocationHelper,
            private val resourceProvider: ResourceProvider,
            private val filterRepository: FilterRepository) : BaseViewModel() {

    var columnCount = ObservableInt(2)
    var refreshing = ObservableBoolean(false)
    var chipList: ObservableList<Chip> = ObservableArrayList()
    var locationChip = ObservableField<Chip>()

    private var animalType = AnimalType.CAT
    private var lastOffset = 0
    private var shouldLoadMoreAnimals = true
    private var firstLoad = true

    private val currentFilter: Observable<Filter>
        get() = filterRepository.currentFilterAndNoFilterIfEmpty
                .doOnSubscribe(this::removeFilterChips)
                .doOnSuccess(this::addFilterChips).toObservable()

    private val currentLocation: Observable<Address>
        get() = locationHelper.fetchCurrentLocation()
                .doOnSubscribe(this::removeLocationChip)
                .doOnNext(this::addLocationChip)

    init {
        dataStore[DiscoverAnimalTypeUseCase::class]?.let {
            animalType = it.animalType
        }
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun updateToolbar() {
        dataStore += DiscoverToolbarTitleUseCase(animalType.toolbarName)
        eventBus += OptionsMenuEvent(this, OptionsMenuState.DISCOVER)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        val filterUpdated = dataStore[FilterUpdatedUseCase::class] != null
        if (firstLoad || filterUpdated) {
            firstLoad = false
            requestPermissionToLoad()
        }
    }

    fun requestPermissionToLoad() {
        if (permissionHelper.hasPermissions(ACCESS_FINE_LOCATION)) {
            loadList()
        } else {
            eventBus += PermissionEvent(this,
                    arrayOf(ACCESS_FINE_LOCATION),
                    locationPermissionResponseListener(),
                    100)
        }
    }

    private fun loadList() {
        listAdapter.clearAnimalItems()
        lastOffset = 0
        shouldLoadMoreAnimals = true
        fetchAnimals()
    }

    fun loadMoreAnimals() {
        if (shouldLoadMoreAnimals) {
            fetchAnimals()
        }
    }

    private fun fetchAnimals() {
        refreshing.set(true)
        dataStore += DiscoverErrorUseCase(null)
        Observable.zip<Address, Filter, FetchAnimalsRequest>(currentLocation,
                currentFilter, BiFunction { location, filter -> FetchAnimalsRequest(animalType, lastOffset, location.postalCode, filter) })
                .flatMap(animalRepository::fetchAnimals)
                .subscribe(this::setAnimalList, this::showError)
                .onLifecycle()
    }

    @Suppress("UNUSED_PARAMETER")
    private fun removeFilterChips(disposable: Disposable) {
        chipList.clear()
    }

    private fun addFilterChips(filter: Filter) {
        filter.apply {
            if (sex !== Sex.MISSING) {
                chipList.add(Chip(resourceProvider.getString(sex.formattedName)))
            }
            if (age !== Age.MISSING) {
                chipList.add(Chip(resourceProvider.getString(age.formattedName)))
            }
            if (size !== AnimalSize.MISSING) {
                chipList.add(Chip(resourceProvider.getString(size.formattedName)))
            }
            if (breed != "") {
                chipList.add(Chip(filter.breed))
            }
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun removeLocationChip(disposable: Disposable) {
        locationChip.set(null)
    }

    private fun addLocationChip(location: Address) {
        locationChip.set(Chip(resourceProvider.getString(R.string.chip_near_location, location.postalCode)))
    }

    private fun showError(throwable: Throwable) {
        refreshing.set(false)
        if (throwable is PetfinderException) {
            dataStore += DiscoverErrorUseCase(throwable.statusCode)
        }
        throwable.printStackTrace()
    }

    private fun setAnimalList(animalListResponse: AnimalListResponse) {
        refreshing.set(false)
        dataStore += DiscoverErrorUseCase(null)
        lastOffset = animalListResponse.lastOffset
        animalListResponse.animalList?.let { animalList ->
            if (animalList.size > 0
                    && listAdapter.animalItems.size > 0
                    && animalList[animalList.size - 1].id == listAdapter.animalItems[listAdapter.animalItems.size - 1].id) {
                shouldLoadMoreAnimals = false
            } else {
                listAdapter.animalItems = animalList.map { animalListItemFactory.newInstance(it) } as ArrayList<AnimalListItemViewModel>
            }
        }
    }

    private fun locationPermissionResponseListener(): PermissionListener {
        return object : PermissionListener {
            override fun onPermissionResult(response: PermissionRequestResponse) {
                if (response.permission == ACCESS_FINE_LOCATION) {
                    if (response.isGranted) {
                        dataStore += DiscoverErrorUseCase(null)
                        requestPermissionToLoad()
                    } else {
                        dataStore += DiscoverErrorUseCase(StatusCode.ERR_NO_LOCATION)
                    }
                }
            }
        }
    }
}
